hack/generator/pkg/astmodel/package_import_set.go

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astmodel
+
+import "sort"
+
+// PackageImportSet represents a set of distinct PackageImport references
+type PackageImportSet struct {
+	imports map[PackageImport]bool
+}
+
+// EmptyPackageImportSet creates a new empty set of PackageImport references
+func EmptyPackageImportSet() *PackageImportSet {
+	return &PackageImportSet{
+		imports: make(map[PackageImport]bool),
+	}
+}
+
+// AddImport ensures the set includes the specified import
+// Adding an import already present in the set is fine.
+func (set *PackageImportSet) AddImport(packageImport PackageImport) {
+	set.imports[packageImport] = true
+}
+
+// AddReference ensures this set includes an import of the specified reference
+// Adding a reference already in the set is fine.
+func (set *PackageImportSet) AddReference(ref PackageReference) {
+	set.AddImport(NewPackageImport(ref))
+}
+
+// Merge ensures that all imports specified in other are included
+func (set *PackageImportSet) Merge(other *PackageImportSet) {
+	for i := range other.imports {
+		set.AddImport(i)
+	}
+}
+
+// Remove ensures the specified item is not present
+// Removing an item not in the set is not an error.
+func (set *PackageImportSet) Remove(packageImport PackageImport) {
+	delete(set.imports, packageImport)
+}
+
+// Contains allows checking to see if an import is included
+func (set *PackageImportSet) ContainsImport(packageImport PackageImport) bool {
+	_, found := set.imports[packageImport]
+	return found
+}
+
+// ImportFor looks up a package reference and returns its import, if any
+func (set *PackageImportSet) ImportFor(ref PackageReference) (PackageImport, bool) {
+	for imp := range set.imports {
+		if imp.PackageReference.Equals(ref) {
+			return imp, true
+		}
+	}
+
+	return PackageImport{}, false
+}
+
+// AsSlice() returns a sorted slice containing all the imports
+func (set *PackageImportSet) AsSlice() []PackageImport {
```

Comment:
There seem to be at least two legitimate orderings to items from this set:
1. Ordered alphabetically including name (if applicable) - which is what this does.
2. Ordered alphabetically ignoring name (i.e. just ordered alphabetically by the `PackageImport`).

I could envision cases where we want all the package references with a name first and all without afterwards too maybe?

I guess I'm saying: Why is this particular ordering blessed as _the_ ordering for getting a slice? What if I wanted a slice in a different order? Is this ordering really intrinsic to the `PackageImportSet`-ness of things or would it make more sense to have this logic defined as a helper in the context of whatever code is trying to render these things into lines in a file?

Reply:
Changed the `AsSlice()` method to return in an undefined order and introduced `AsSortedSlice()` for when the order matters.

Reply:
Nice.

Reply:
Actually, it's not an undefined order right now - is that intentional (so that it's at least repeatable), or an oversight? See my other comment below about double-sorting as well.

Reply:
An oversight on my part - should have removed the sorting from `AsSlice()` but I got sidetracked doing the testing on `AsSortedSlice()`.